### PR TITLE
[Feature] Allow use of workflows inside of workflows

### PR DESCRIPTION
**Pull request by AlanCoding**
Connect https://github.com/ansible/awx/issues/2252

### CODE REVIEW

File: awx/main/models/workflow.py [hunk 1]
```diff
@@ -472,6 +480,16 @@ def notification_data(self):
...
 
+    def get_ancestor_workflows(self):
+        ancestors = []
+        wj = self
+        while True:
+            wj = wj.get_workflow_job()
+            if (not wj) or (not wj.workflow_job_template_id):
+                break
+            ancestors.append(wj.workflow_job_template_id)
+        return ancestors
```
Comment (AlanCoding) [inline awx/main/models/workflow.py:532]: might want to track a set of workflow jobs, so that we can bail if we find that this loop is repeating itself. The tracking of the templates will not provide that, and right now it is vulnerable to an infinite loop if someone created cycles via the shell.

File: awx/main/models/workflow.py [hunk 2]
```diff
@@ -472,6 +480,24 @@ def notification_data(self):
     def task_impact(self):
         return 0
 
+    def get_ancestor_workflows(self):
```
Comment (chrismeyersfsu) [inline awx/main/models/workflow.py:516]: Hot

**Comment by AlanCoding** [comment]
Development has now finished on this, and we're getting into pre-merge activities. Taking off the WIP from the title because of that.

File: docs/workflow.md [hunk 3]
```diff
@@ -26,6 +26,18 @@ See the document on saved launch configurations for how these are processed
 when the job is launched, and the API validation involved in building
 the launch configurations on workflow nodes.
 
+#### Workflows as Workflow Nodes
+
+A workflow can be added as a node in another workflow. The child workflow is the associated
+`unified_job_template` that the node references, when that node is added to the parent workflow.
+When the parent workflow dispatches that node, then the child workflow will begin running, and
+the parent will resume execution of that branch when the child workflow finishes.
+Branching into success / failed pathways is decided based on the status of the child workflow.
```
Comment (jladdjr) [inline docs/workflow.md:35]: @AlanCoding (cc @Spredzy) - When I worked with the original implementation of workflows, I remember that it was pretty much impossible to get them into a failure state (I never could). Has the definition of workflow success / failure / error changed since workflows were originally written? Wondering if it's still basically impossible for workflows to fail.

Comment (AlanCoding) [reply docs/workflow.md:35]: Yes, some time ago jangsutsr implemented workflow failure condition. See https://github.com/ansible/awx/pull/382

The history of workflow status was this:
 - pre-release in initial implementation we had the "completed" state, so they would either be completed or error
 - various ideas were floated around about user-defined failure, but nothing was agreed on, except that we probably wanted this notion in the future, and a result of this was that the (blue) completed state was actually walked back, the implementation because either "success" or "error"
 - some time after that, the failure condition was implemented, so it became "success", "failure", or "error"

You could also add "cancel" to these. I have some pending patches for bugs with the cancel status too.

Because of the current definition of workflow failure, the failure scenario for sliced jobs just _happens_ to already be identical to if they were not sliced, and workflow jobs follow the same pattern too. For sliced jobs, all nodes are leaf nodes, so failure of any fails the workflow, just as any failure on a host fails the playbook. For workflows, it's the leaf nodes.

In other words, the branching of the network with workflows-in-workflows required no implementation, and is taken to already work with no change to existing logic. Sure, test it anyway, if you see fit.

File: awx/ui/client/src/templates/main.js [hunk 4]
```diff
@@ -409,17 +409,15 @@ angular.module('templates', [surveyMaker.name, jobTemplates.name, labels.name, p
...
                     data: {
                         formChildState: true
                     },
                     params: {
                         job_template_search: {
                             value: {
                                 page_size: '5',
-                                order_by: 'name'
+                                order_by: 'name',
+                                role_level: 'execute_role'
```
Comment (mabashian) [inline awx/ui/client/src/templates/main.js]: Nice!

File: awx/ui/client/src/templates/main.js [hunk 5]
```diff
@@ -708,15 +706,18 @@ angular.module('templates', [surveyMaker.name, jobTemplates.name, labels.name, p
                                 delete list.fields.labels;
                                 delete list.fieldActions;
                                 list.fields.name.columnClass = "col-md-8";
+                                list.fields.name.tag = i18n._('WORKFLOW');
+                                list.fields.name.showTag = "{{job_template.type === 'workflow_job_template'}}";
```
Comment (mabashian) [inline awx/ui/client/src/templates/main.js]: I wonder if the `job_template` variable needs to be changed to just `template`.  I think we could achieve this by changing list.iterator to `template` and list.name to `templates`.  We'd have to update all the other places that that might impact.

**Comment by mabashian** [comment]
I'm wondering if we should hide this button for workflow job templates in the list:

screen shot 2018-10-26 at 8 10 35 pm

It doesn't really display any relevant information anyway:

screen shot 2018-10-26 at 8 10 49 pm

**Comment by mabashian** [comment]
When viewing a running workflow it looks like the key is running of the page with the addition of the new Workflow Sync icon/description.  Ping @trahman73 we might need to come up with a more responsive solution here:

screen shot 2018-10-26 at 8 14 40 pm

**Comment by trahman73** [comment]
@mabashian We can use the same Key from the WF Editor view here:

screen shot 2018-10-26 at 8 44 04 pm

https://tower-mockups.testing.ansible.com/mockups/workflows/workflows-key/

The resulting WF View screen toolbar would look like this:
https://tower-mockups.testing.ansible.com/mockups/jobs/jobs-workflow/

**Comment by mabashian** [comment]
@marshmalien I think we just need to change `job_template` to `template` in https://github.com/ansible/awx/blob/workflows_squared/awx/ui/client/src/partials/job-template-details.html.  This icon:

screen shot 2018-11-01 at 8 56 16 pm

is hidden for job templates in that list.  

It also looks like the On Failure label is missing in the workflow results key:

screen shot 2018-11-01 at 8 59 10 pm

Once we get those two things fixed I think we'll be good to go

**Comment by mabashian** [comment]
UI portion lgtm!

**Comment by marshmalien** [comment]
Connect: https://github.com/ansible/awx/issues/2581 https://github.com/ansible/awx/issues/2582

**Comment by AlanCoding** [comment]
My last commit was due to feedback from @Spredzy that the error text was very hard for a user to digest in the event of recursion. I have tried to patch that up:

screen shot 2018-11-05 at 2 50 24 pm

**Comment by marshmalien** [comment]
Connect: #2617 #2584

File: awx/main/models/workflow.py [hunk 6]
```diff
@@ -505,6 +513,24 @@ def notification_data(self):
...
+        """Returns a list of WFJTs that are indirect parents of this workflow job
+        say WFJTs are set up to spawn in order of A->B->C, and this workflow job
+        came from C, then C is the parent and [B, A] will be returned from this.
+        """
+        ancestors = []
+        wj_ids = set([self.pk])
+        wj = self.get_workflow_job()
+        while wj and wj.workflow_job_template_id:
+            if wj.pk in wj_ids:
+                logger.critical('Cycles detected in the workflow jobs graph, '
```
Comment (chrismeyersfsu) [inline awx/main/models/workflow.py:526]: Print the id of wfj and wfjt where the cycle is detected.

File: awx/main/scheduler/task_manager.py [hunk 7]
```diff
@@ -165,7 +181,7 @@ def process_finished_workflow_jobs(self, workflow_jobs):
                 workflow_job.status = new_status
                 workflow_job.start_args = ''  # blank field to remove encrypted passwords
                 workflow_job.save(update_fields=['status', 'start_args'])
-                connection.on_commit(lambda: workflow_job.websocket_emit_status(workflow_job.status))
+                workflow_job.websocket_emit_status(workflow_job.status)
```
Comment (chrismeyersfsu) [inline awx/main/scheduler/task_manager.py:184]: websocket_emit_status has connection.on_commit in it :+1:

**Review by chrismeyersfsu — APPROVED** [review]

**Comment by mabashian** [comment]
UI changes look good

**Review by mabashian — APPROVED** [review]

File: awx/main/models/workflow.py (hunk 6 above)
Comment (ryanpetrello) [inline awx/main/models/workflow.py:526]: 👍

**Review by ryanpetrello — APPROVED** [review]
